homeassistant/components/tesla/__init__.py

```diff
@@ -34,69 +40,150 @@
     extra=vol.ALLOW_EXTRA,
 )
 
-NOTIFICATION_ID = "tesla_integration_notification"
-NOTIFICATION_TITLE = "Tesla integration setup"
+
+@callback
+def _async_save_refresh_token(hass, config_entry, access_token, token):
+    hass.config_entries.async_update_entry(
+        config_entry,
+        data={**config_entry.data, CONF_ACCESS_TOKEN: access_token, CONF_TOKEN: token},
+    )
 
 
 async def async_setup(hass, base_config):
     """Set up of Tesla component."""
-    config = base_config.get(DOMAIN)
 
+    def _update_entry(email, scan_interval=300, data=None, options=None):
+        data = data or {}
+        options = options or {CONF_SCAN_INTERVAL: scan_interval}
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if email == entry.title:
+                hass.config_entries.async_update_entry(
+                    entry, data=data, options=options
+                )
+
+    config = base_config.get(DOMAIN)
+    if not config:
+        return True
     email = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
-    update_interval = config.get(CONF_SCAN_INTERVAL)
-    if hass.data.get(DOMAIN) is None:
+    scan_interval = config.get(CONF_SCAN_INTERVAL, 300)
+    if email in configured_instances(hass):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            if email == entry.title:
+                try:
+                    websession = aiohttp_client.async_get_clientsession(hass)
+                    controller = teslaAPI(
+                        websession, email=email, password=password, update_interval=300
+                    )
+                    (refresh_token, access_token) = await controller.connect(
+                        test_login=True
+                    )
+                    hass.config_entries.async_update_entry(
+                        entry,
+                        data={
+                            CONF_ACCESS_TOKEN: access_token,
+                            CONF_TOKEN: refresh_token,
+                        },
+                        options={CONF_SCAN_INTERVAL: scan_interval},
+                    )
+                except TeslaException as ex:
+                    _LOGGER.warning(
+                        "Configuration.yaml credentials cannot communicate with Tesla API: %s",
+                        ex.message,
+                    )
+                break
+    else:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data={CONF_USERNAME: email, CONF_PASSWORD: password},
+            )
+        )
+        async_call_later(hass, 15, lambda _: _update_entry(email, scan_interval))
```

Comment:
what does this to ? Also lambda is called in sync context and _update_entry calls async methods.

Reply:
This function updates the config entry's options. Without the delay, it wasn't finding an entry to update.